### ISSUE TITLE
Check the skills needed in the challenges

### ISSUE THREAD

**Issue by fmichonneau** [datacarpentry/R-ecology-lesson#189]
make sure that all the skills needed to answer a challenge question have been covered previously in the lessons

- [x] ~~Before we start~~ (no challenge in this lesson)
- [x] Introduction to R
- [x] Starting with data
- [x] Aggregating and analyzing data with dplyr
- [x] Data visualization with ggplot2
- [x] R and databases

**Comment by sdtaylor** [comment]
## Into to R challenges
### Creating objects
- create new objects using numbers and other objects.   
Covered previously? Yes

### Vectors and data types
- Make vectors of mixed types and decide what the final vector type will be.
Covered previously? No. Some students likely think about trying this but this challenge is the first time it's presented. The use of `class()` is covered for them to explore it on their own, and an explanation is included. I don't suggest changing anything.  
- Why does `"four" > "five"` return `TRUE`?
**Covered  previously? No.  Logical tests are in the next section.**   

### Subsetting vectors  
No challenges

### Missing Data  
- remove NA's from a vector
Covered previously? Yes, but a more explicit explanation of what `!` does would probably help.  
- Calculate the median from a vector with NA's in it.  
Covered previously? Yes. `na.rm` was shown as an option in `mean()` and `max()` so a student should be able to conclude that `median()` has this option too.

**Comment by sdtaylor** [comment]
## Starting with data challenges
### What are data frames?  
No challenges

### Inspecting data.frame objects  
- 3 questions on the `surveys` data.frame based on the output of `str(surveys)`  
Covered previously? Yes. The dimensions and class of data.frame objects are looked at just before this. Number of species in based on the factor levels of that column, which are covered in the prior chapter.   

### Indexing and subsetting data frames
- Create a dataframe of the first 200 rows
Covered previously? yes
- Extract the last row of the dataframe using `nrow()`  
Covered  previously? Yes. But students will need to make the connection that `nrow()` outputs a number which can be used in subsetting.  
- Use `nrow()` to get the last 100 rows.  
**Covered previously? Yes/No.** Students will have to use `nrow()` twice but otherwise this is the same as above. The first way I would try this is  `surveys[nrow(surveys)-100:nrow(surveys),]`, but this doesn't work as it evaluates `100:nrow(surveys)` first and then subtracts that from `nrow(surveys)`. To fix that there must be  `surveys[(nrow(surveys)-100):nrow(surveys),]`, which probably won't be obvious to students. Is there another answer that was intended here?  

### Factors
- rename the level names inside the `sex` factor.
Covered previously? Yes  
- change order of levels in `sex` for plotting.
Covered previously? Yes

- Find mistakes in small data frame creation.
Covered previously? Yes. Mistakes are commas and equal vector lengths. 
- Confirm class types in a small data frame.
Covered previously? Yes. The data.frame has issues with mixed data types which was covered in "Intro to R"

**Comment by sdtaylor** [comment]
## Data manipulation using dplyr challenges

- Subset the survey data to a specific year and 3 columns.
covered previously? Yes. Requires `filter()` and `select()`, both already covered. 

- Create a dataframe with specific requirements. 
Covered  previously? Yes. Requires `filter()`, `mutate()`, `select()`, and `is.na()`, all covered previously. 

- Count the number of individuals caught in each plot type.
Covered previously? Yes. Requires `group_by()` and `tally()`

- Get statistics on foot length for each species.
Covered previously? Yes. Will also require using `na.rm`, which has been covered. 

- Get heaviest animal in each year and return specific columns.  
Covered previously? Yes. A little confusing though:
> What was the heaviest animal measured in each year? Return the columns `year`, `genus`, `species_id`, and `weight`.  

To get the absolute heaviest animal you must `group_by(year)`, but this will not return the `genus` and `species_id` columns. Those could be included in `group_by()` but then you'd have to manually filter to find the heaviest. `top_n()` would work here but is not introduced yet.   

- Use `group_by()`, `summarize()` and `n()` instead of `tally()`.
Covered previously? No.  This is the first mention of `n()`, but it suggests to look at the help for it, which is a good exercise.

**Comment by sdtaylor** [comment]
## Visualization with ggplot2
- Describe why you would use `geom_hex()` for dense data.
Previously covered? Yes

### Boxplot
- Create new plots based on a different `geom_` function.
Covered previously? Yes.
- Add color the one of the boxplots based on species.
Covered previously? Yes. 

### Faceting and themes
- Create a yearly timeseries plot with the average weight of each species.
Previously covered? Yes. Students must create a new dataframe with `group_by()`, covered in the last chapter. 

**This chapter is good to go.**

**Comment by sdtaylor** [comment]
## R and databases
### Complex database queries
- write a query that summarizes total rodents by plot and year.
Previously covered? Yes. Requires a `group_by()` learned previously.

- Summarize counts of genus by `plot_id`
**Previously covered? No. If I read this correctly then it would require `n_distinct()` which hasn't been covered yet.** 

- Add the final table to the DB and re-run prior commands.
Previously covered? Yes

**Comment by fmichonneau** [comment]
Thanks so much for your work on this @sdtaylor !

It's great, and it will be useful down the road as we modify the lessons.

I edited your comments to include issues that I open to reflect the 3 issues your raised:
- #197
- #198
- #200 

so it will be easier to discuss resolution and refer to in pull requests.

**Comment by fmichonneau** [comment]
Thanks again for doing this. I think all the issues you've raised have now been addressed. Closing.

**Comment by sdtaylor** [comment]
:+1:
